packages/hardhat-core/src/internal/hardhat-network/provider/vm/vm-adapter.ts

```diff
@@ -1,17 +1,38 @@
-import type {
-  EVMResult,
-  InterpreterStep,
-  Message,
-} from "@nomicfoundation/ethereumjs-evm";
 import type { RunTxResult } from "@nomicfoundation/ethereumjs-vm";
 import type { Block } from "@nomicfoundation/ethereumjs-block";
 import type { TypedTransaction } from "@nomicfoundation/ethereumjs-tx";
 import type { Account, Address } from "@nomicfoundation/ethereumjs-util";
 import type { RpcDebugTracingConfig } from "../../../core/jsonrpc/types/input/debugTraceTransaction";
 import type { RpcDebugTraceOutput } from "../output";
+import { EvmError } from "@nomicfoundation/ethereumjs-evm";
 
 export type Trace = any;
 
+export interface TracingMessage {
+  to?: Address;
+  depth: number;
+  data: Buffer;
+  value: bigint;
+  codeAddress: Address;
+}
+export interface TracingStep {
+  pc: number;
+}
+export interface TracingMessageResult {
+  createdAddress?: Address;
+  execResult: {
+    exceptionError?: EvmError;
+    returnValue: Buffer;
+    executionGasUsed: bigint;
+  };
+}
+
+export interface TracingCallbacks {
+  beforeMessage: (message: TracingMessage, next: any) => Promise<void>;
+  step: (step: TracingStep, next: any) => Promise<void>;
+  afterMessage: (result: TracingMessageResult, next: any) => Promise<void>;
+}
+
```

Comment:
@Wodann this is the relevant part of this change. These callbacks should be called with those values.

We need to figure out if the `next` callbacks will be called from rethnet or from the adapter, but that's a minor detail.

Reply:
Would you be opposed to defining these in `rethnet_evm_nqpi`? Then they are the same between Rethnet and Hardhat automatically

Reply:
No problem at all!

Reply:
Done :slightly_smiling_face: